cpp/src/decisiontree/decisiontree.h

```diff
@@ -81,75 +81,102 @@ struct DecisionTreeParams {
 	void fit(const ML::cumlHandle& handle, T *data, const int ncols, const int nrows, int *labels, unsigned int *rowids,
 			const int n_sampled_rows, const int unique_labels, DecisionTreeParams tree_params);
 
-	/* Predict labels for n_rows rows, with n_cols features each, for a given tree. rows in row-major format. */
-	void predict(const ML::cumlHandle& handle, const T * rows, const int n_rows, const int n_cols, int* predictions, bool verbose=false) const;
-
-	// Printing utility for high level tree info.
-	void print_tree_summary() const;
-
-	// Printing utility for debug and looking at nodes and leaves.
-	void print() const;
-
 private:
-	// Same as above fit, but planting is better for a tree then fitting.
-	void plant(const cumlHandle_impl& handle, T *data, const int ncols, const int nrows, int *labels, unsigned int *rowids, const int n_sampled_rows, int unique_labels,
-		   int maxdepth = -1, int max_leaf_nodes = -1, const float colper = 1.0, int n_bins = 8, int split_algo_flag = SPLIT_ALGO::HIST, int cfg_min_rows_per_node=2, bool cfg_bootstrap_features=false);
+	/* depth is used to distinguish between root and other tree nodes for computations */
+	void find_best_fruit_all(T *data, int *labels, const float colper, MetricQuestion<T> & ques, float& gain, unsigned int* rowids,
+							const int n_sampled_rows, MetricInfo<T> split_info[3], int depth);
+}; // End DecisionTreeClassifier Class
 
-	TreeNode<T> * grow_tree(T *data, const float colper, int *labels, int depth, unsigned int* rowids, const int n_sampled_rows, GiniInfo prev_split_info);
+template<class T>
+class DecisionTreeRegressor : public DecisionTreeBase<T, T> {
+public:
+	void fit(const ML::cumlHandle& handle, T *data, const int ncols, const int nrows, T *labels, unsigned int *rowids,
+			const int n_sampled_rows, DecisionTreeParams tree_params);
 
+private:
 	/* depth is used to distinguish between root and other tree nodes for computations */
-	void find_best_fruit_all(T *data, int *labels, const float colper, GiniQuestion<T> & ques, float& gain, unsigned int* rowids,
-							const int n_sampled_rows, GiniInfo split_info[3], int depth);
-	void split_branch(T *data, GiniQuestion<T> & ques, const int n_sampled_rows, int& nrowsleft, int& nrowsright, unsigned int* rowids);
-	void classify_all(const T * rows, const int n_rows, const int n_cols, int* preds, bool verbose=false) const;
-	int classify(const T * row, const TreeNode<T> * const node, bool verbose=false) const;
-	void print_node(const std::string& prefix, const TreeNode<T>* const node, bool isLeft) const;
-}; // End DecisionTree Class
+	void find_best_fruit_all(T *data, T *labels, const float colper, MetricQuestion<T> & ques, float& gain, unsigned int* rowids,
+							const int n_sampled_rows, MetricInfo<T> split_info[3], int depth);
+}; // End DecisionTreeRegressor Class
 
 } //End namespace DecisionTree
 
 
 // Stateless API functions
+
```

Comment:
There is one very interesting issue that just arose while @Salonijain27 was writing cython classes for Random Forest: this are *not* stateless APIs! This functions seem to just be an alternative way of calling the `DecisionTreeClassifier` (or equivalent RF class) Fit, but the object is still stateful! A stateless API means that the state is kept by the client (python) and the C++ code does not keep the state, in this case the C++ object still keeps the state in spite of there being a `fit` function outside of it. 

This not only complicates the cython work (not a deal breaker and it can be done to work) but also potentially complicates the pickling (model saving) functionality. For all the algorithms that are stateless (most of cuML) we automatically get that functionality since the python objects are the state of the algorithm. Wanted to start a quick conversation regarding this.

Reply:
(Of course feel free to correct any of this if the analysis/conclusions above are wrong!)

Reply:
Hmm.. that's interesting. IIUC, UMAP/kNN both expose such "stateful" classes, no? How are they being handled in cython world?

@cjnolet ^^^

Reply:
Yes they do, and the reason UMAP and KNN are the 2 classes that *cannot* be pickled currently (this was raised only a few weeks ago when some user ran into that issue). There is even an open issue #415 to create a stateless version of knn

Reply:
Also this is not necessarily something that needs to change per se, though it could make things much easier. We can look into using the `__getstate__` and `__setstate__` python functions to implement the pickling. Will be looking into this today and the first days of next week. But even then the question of whether we want a uniformity in stateless(or fulness) of the C++ API still stands

Reply:
I did think it was strange in the initial PR for the RF and DT algorithms that the stateless functions are being called by functions that maintain their own state and not the other way around. Unfortunately, as we were ramping up to release, I did not have the time to make the point before it was merged. I did and still do have concerns about it, however. UMAP was implemented with a stateful convenience API built on top of a set of stateless functions and converting the cython to use those stateless functions will be trivial.

KNN was originally forced to be stateful because the FAISS Indexing API was designed in such a way that wrapping it with cython was non-trivial and the state was better off hidden from the cython layer altogether. Recent updates to their API have eliminated this problem and kNN will soon be reduced to a simple stateless `kneighbors` function on the C++ side. I'm also considering moving it to the prims.

I do think we should continue to provide stateful convenience classes in the C++ layer. I would recommend, as a solution here, to expose a set of flat stateless functions, have the caller maintain the state, and also expose a stateful convenience API that is based on those stateless functions, maintaining the state for the users of the C++ API.

Reply:
We plan to implement a newer version of flat API for 0.9 release. The current rfRegressor API is inline to branch-0.8 rfClassifier